### PR TITLE
dr::mhp::sort

### PR DESCRIPTION
**Pull request by mateuszpn**
mhp::sort() implementing distributed bucket sort, with data redistribution to match segment sizes

### CODE REVIEW

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
I've only reviewed `communicator` and stopped.
@mateuszpn , please rerequest review from me once compilation is fixed (all checks are green) and comments to `communicator` are applied, so I will continue with reading `sort` algorithm.

File: include/dr/detail/communicator.hpp [hunk 1]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+
+    MPI_Alltoallv(sendbuf, _sendcnt.data(), _senddsp.data(), MPI_BYTE, recvbuf,
+                  _recvcnt.data(), _recvdsp.data(), MPI_BYTE, MPI_COMM_WORLD);
+  }
+
+  template <rng::contiguous_range R>
+  void alltoallv(R &s, std::vector<int> &sendcnt, std::vector<int> &senddsp,
+                 R &r, std::vector<int> &recvcnt, std::vector<int> &recvdsp) {
+
+    alltoallv(s.data(), sendcnt.data(), senddsp.data(), r.data(),
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:153]: 1. General remark - the changes in communicator are moved to PR #391. I decided to separate the PRs as they relate to different problems, so every change (also coming from comments here) is now made in #391
2. ranges::contiguous_range must have ranges::data(r) implemented and typically it does it with data() method

File: include/dr/detail/communicator.hpp [hunk 2]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+  template <rng::contiguous_range R>
+  void alltoall(R &sendr, R &recvr, int size) {
+    alltoall(sendr.data(), recvr.data(), size);
+  }
+
+  template <typename T>
+  void alltoallv(T *sendbuf, int *sendcounts, int *sdispls, T *recvbuf,
+                 int *recvcounts, int *rdispls) {
+
+    std::span<int> sendcnt(sendcounts, sendcounts + size_);
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:126]: Thanks, will do it in #

File: include/dr/detail/communicator.hpp [hunk 3]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+    std::span<int> senddsp(sdispls, sdispls + size_);
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:127]: 1. I a pointer is passed, there is no way to verify the size of array
2. The convention in this library is to have two functions covering single MPI call - one with inteface very similar to original one (with pointers), second with ranges. I follow this approach

File: include/dr/detail/communicator.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/detail/communicator.hpp:153]: right, so use of `rng::data(s)` is correct, but `s.data()` may be not correct. Please change to `rng::data(s)` (and `rng::size()` in some other place)

File: include/dr/detail/communicator.hpp [hunk 4]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+                 count * sizeof(T), MPI_BYTE, mpi_comm_);
+  }
+
+  template <rng::contiguous_range R>
+  void alltoall(R &sendr, R &recvr, int size) {
+    alltoall(sendr.data(), recvr.data(), size);
+  }
+
+  template <typename T>
+  void alltoallv(T *sendbuf, int *sendcounts, int *sdispls, T *recvbuf,
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:123]: This is called by alltoallv() overload (lines 149+). Two calls due to convention taken by creators of the class

File: include/dr/detail/communicator.hpp [hunk 5]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
               request);
   }
 
   template <rng::contiguous_range R>
   void irecv(R &data, std::size_t source, tag t, MPI_Request *request) const {
-    irecv(data.data(), data.size(), source, int(t), request);
+    irecv(data.data(), data.size(), source, t, request);
+  }
+
+  template <typename T> void alltoall(T *sendbuf, void *recvbuf, int count) {
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:112]: Yes, during the work I realized I should split the changes in 2 PRs - because they refer to different areas of library

File: include/dr/detail/communicator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/detail/communicator.hpp:127]: > I a pointer is passed, there is no way to verify the size of array

I meant assertion in the functoin where you have vectors.

> The convention in this library is to have two functions covering single MPI call

Indeed, I see there is such a convention. Someday, and that day may never come, I will remove half of useless functions in communicator. But until that day, let's have two alltoall-s.

File: include/dr/detail/communicator.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/detail/communicator.hpp:123]: Is that `alltoallv()` overload called somewhere except tests?

File: include/dr/mhp/algorithms/sort.hpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+#include <dr/detail/logger.hpp>
+#include <dr/detail/onedpl_direct_iterator.hpp>
+#include <dr/detail/ranges_shim.hpp>
+#include <dr/mhp/global.hpp>
+
+namespace dr::mhp {
+
+template <dr::distributed_range R, typename Compare = std::less<>>
+void sort(R &r, Compare comp = Compare()) {
+  using T = R::value_type;
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:26]: We use T as an element type in most cases, but no difference for me

[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:26]: Not used anymore

[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:26]: Typically we use T as simple (arithmetic, non-range) element type, and non-range template argument type, too, still no difference for me.

File: include/dr/detail/communicator.hpp [hunk 7]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+    MPI_Isend(data, size * sizeof(T), MPI_BYTE, dest, int(t), mpi_comm_,
               request);
   }
 
   template <rng::contiguous_range R>
-  void isend(const R &data, std::size_t source, tag t,
+  void isend(const R &data, std::size_t dest, tag t,
              MPI_Request *request) const {
-    isend(data.data(), data.size(), source, int(t), request);
+    isend(data.data(), data.size(), dest, t, request);
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: contiguous_range has ranges::data(r), typically implemented as .data() - do you know other implementation? Still, to be conformant with specification, changed to rng::data(r) in all places where the type is general (vector and span provide data())
AFAIK size() is available to every range (not in all cases size() has constant time of execution)

File: include/dr/detail/communicator.hpp [hunk 8]
```diff
@@ -85,28 +85,82 @@ class communicator {
   }
 
   template <typename T>
-  void isend(const T *data, std::size_t size, std::size_t source, tag t,
+  void isend(const T *data, std::size_t size, std::size_t dest, tag t,
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: Ranks are always 0 or positive values, so following general approach in the library I changed all ranks to size_t

File: include/dr/detail/communicator.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: This overload of alltoall is not used and removed

File: include/dr/detail/communicator.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: changed to rng::data(r) and rng::size(r) in all occurrences of .data() and .size()

File: include/dr/detail/communicator.hpp [hunk 9]
```diff
@@ -85,28 +85,82 @@ class communicator {
...
+  }
+
+  template <typename T> void alltoall(T *sendbuf, void *recvbuf, int count) {
+    MPI_Alltoall(sendbuf, count * sizeof(T), MPI_BYTE, recvbuf,
+                 count * sizeof(T), MPI_BYTE, mpi_comm_);
+  }
+
+  template <rng::contiguous_range R>
+  void alltoall(R &sendr, R &recvr, int size) {
+    alltoall(sendr.data(), recvr.data(), size);
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: 1) is discussed in other place 2) and 3) is done

File: include/dr/detail/communicator.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp]: Its done

File: test/gtest/mhp/sort.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    std::sort(l_v.begin(), l_v.end());
+    dr::mhp::sort(d_v);
+
+    barrier();
+    EXPECT_TRUE(equal(l_v, d_v));
+  }
+}
+
+TEST(MhpSort, Sort_reverse) {
```
Comment (lslusarczyk) [inline test/gtest/mhp/sort.cpp:34]: This test is a copy of above one.
loops/logic are not warmly welcomed in perfect unit-tests too
Copy-pastes in tests are more allowed as tests has to be readable, but here it is just one param.
I would create helper func `test_sort_randomly(size, functor)` and make your tests look sth like this
```
TEST(MhpSort, random_sort)
{
test_sort_randomly(1, std::less<T>());
test_sort_randomly(comm_size-1, std::less<T>());
test_sort_randomly(4, std::less<T>());
test_sort_randomly(10, std::less<T>());
}

TEST(MhpSort, random_reverse_sort)
{
test_sort_randomly(1, std::greater<T>());
test_sort_randomly((comm_size-1)*(comm_size-1), std::greater<T>());
test_sort_randomly(7, std::greater<T>());
test_sort_randomly(23, std::greater<T>());
}
```

File: test/gtest/mhp/sort.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    dr::mhp::barrier();
+
+    dr::mhp::sort(d_v, std::greater<T>());
+    std::sort(l_v.begin(), l_v.end(), std::greater<T>());
+    barrier();
+
+    EXPECT_TRUE(equal(l_v, d_v));
+  }
+}
```
Comment (lslusarczyk) [inline test/gtest/mhp/sort.cpp:54]: I think non-random sort tests are also needed. Please put them into fixed 3-ranks suite and create tests with predefined input, e.g:
[1,1,1,1,1,1,1,99,1,1,1] - many_small_values_one_greater
[10,15,13,19,  55, 53, 59, 51,  33, 31, 39, 34] - three_distinct_groups_test
[1,2,3,4,5,6] - sorted_input
[6,5,4,3,2,1] - reversed_sorted_input
[1,2,1,2,1,2,1,2,1,2,1,2] - ....
[1] - just one value
[] - no values
[1,2] - two values
[3,1,2] - three values
[3,1,2,4] - four values
float type test
[1,1,1,1,2,2,2,2] - sorted_two_val_only
[2,2,2,2,1,1,1,1] - reverse_sorted_two_val_only
[1,2,3,4,5,4,3,2,1] - pyramid_test
[5,4,3,2,1,2,3,4,5] - pyramid_reverse_test
[0,0,0,0,0,0,0,0,0] - same_value_test
[1, 2,... 1234567] million_sorted_test
[1234567, 1234566, ... 1] million_reverse_sorted_test
[...] big random test, simply: `{ test_sort_randomly(9876543, std::less<T>()); }`

Any other test cases which you think sort algorithm goes into corner cases, works badly, handles some rare `if`-s...

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
Tests reviewed. I've suggested some test cases. Before I will go into reviewing code I'd like to be sure they do pass.

File: test/gtest/mhp/sort.cpp (hunk 11 above)
Comment (mateuszpn) [reply test/gtest/mhp/sort.cpp:77]: 0 to 4 elements tests are already in random tests (tests of size comm_size - 1 ). I think it is enough.
The other tests are added.

File: test/gtest/mhp/sort.cpp (hunk 10 above)
Comment (mateuszpn) [reply test/gtest/mhp/sort.cpp]: Redundant elements moved to helper fuctions. Tests for std::less and std::greater can be automatic, no need to list them separately.

File: include/dr/detail/communicator.hpp [hunk 12]
```diff
@@ -124,12 +124,12 @@ class communicator {
...
-  template <rng::contiguous_range R>
-  void alltoallv(const R &sendbuf, const std::vector<std::size_t> &sendcnt,
-                 const std::vector<std::size_t> &senddsp, R &recvbuf,
+  template <rng::contiguous_range R1, rng::contiguous_range R2>
+  void alltoallv(const R1 &sendbuf, const std::vector<std::size_t> &sendcnt,
+                 const std::vector<std::size_t> &senddsp, R2 &recvbuf,
                  const std::vector<std::size_t> &recvcnt,
                  const std::vector<std::size_t> &recvdsp) {
-    using valT = typename R::value_type;
+    using valT = typename R2::value_type;
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:132]: Containers can be of different types, but the elements stored must be of the same type. However, there is no way to check this because not all containers implement value_type (ranges::subrange<int *> in our case).

[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/detail/communicator.hpp]: At least [decltype](https://en.cppreference.com/w/cpp/language/decltype) could be used to implement the check. But there is a nicer way: because not all containers implement value_type there is  [rng::range_value_t](https://en.cppreference.com/w/cpp/ranges/iterator_t) which should always work.

File: test/gtest/common/sort.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  std::sort(v.begin(), v.end(), func);
+  xhp::sort(d_v, func);
+
+  barrier();
+  EXPECT_TRUE(equal(v, d_v));
+}
+
+void test_sort2s(LV &v) {
+  test_sort(v, std::less<T>());
+  test_sort(v, std::greater<T>());
```
Comment (lslusarczyk) [inline test/gtest/common/sort.cpp:29]: `test_sort` modifies `v`, therefore the second `test_sort` always gets sorted data.
Please take `LV` by value in `test_sort`

File: test/gtest/common/sort.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  test_sort_randomvec(17);
+  test_sort_randomvec(123);
+}
+
+TEST(Sort, BigRandom) { test_sort_randomvec(1234567, 65535); }
+
+TEST(Sort, NonRandom) {
+  LV v;
+  v = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
+  test_sort2s(v);
```
Comment (lslusarczyk) [inline test/gtest/common/sort.cpp:50]: maybe simply `test_sort2s({1, 1, 1, 1, 1, 1, 1, 1, 1, 1})`?

File: test/gtest/common/sort.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  test_sort_randomvec(1);
+  test_sort_randomvec(comm_size - 1);
+  test_sort_randomvec((comm_size - 1) * (comm_size - 1));
+  test_sort_randomvec(17);
+  test_sort_randomvec(123);
+}
+
+TEST(Sort, BigRandom) { test_sort_randomvec(1234567, 65535); }
+
+TEST(Sort, NonRandom) {
```
Comment (lslusarczyk) [inline test/gtest/common/sort.cpp:47]: putting each case into separate `TEST()` will have two benefits:
- you will see faster which test case fails
- you will execute all other tests even if one of them fails
Disadvatage is that one has to invent names for tests, but this may be also an improvement for readability,

File: test/gtest/common/sort.cpp (hunk 14 above)
Comment (mateuszpn) [reply test/gtest/common/sort.cpp:50]: test_sort2s takes arg by reference, it would be too many copies if I changed this, too

Comment (lslusarczyk) [reply test/gtest/common/sort.cpp:50]: It will not. It's a just a unit test. I guess nobody will even be able to measure time difference if you will take things by value in all places.

In unit tests simplicity is so important (reader doesn't need to think who modifies object, test is not polluted by `const &`, `&`, `&&` and similar things), that taking by value should be default way of writing functions, unless one _observes_ that taking by ref really speeds up the test which is too long. Then we should change by-valu to by-ref.

Comment (mateuszpn) [reply test/gtest/common/sort.cpp:50]: Removing big_random test you suggested then. It took too long time anyway, and it was not clear what it tested, comparing to other tests.

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
comments from first part of review, still reviewing

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
second part of comments, next will come tomorrow

File: include/dr/mhp/algorithms/sort.hpp [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+#include <dr/mhp/global.hpp>
+
+namespace dr::mhp {
+
+namespace __detail {
+
+template <rng::forward_range R, typename Compare>
+void local_sort(R &r, Compare &&comp) {
+  if (rng::size(r) >= 2) {
+#ifdef SYCL_LANGUAGE_VERSION
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:32]: Isn't it redundant? use_sycl is set in global.hpp:37 within the same #ifdef

[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/mhp/algorithms/sort.hpp:32]: it is not, init is not always called, see e.g. `dr_init()` in mhp-tests.cpp, by default global var is false if init is not called, see line: `bool use_sycl_ = false;`

(fix comment you wrote in exclusive_scan about it)

File: include/dr/mhp/algorithms/sort.hpp [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+
+  /* send and receive data belonging to each node */
+
+  std::vector<valT> vec_recvdata(_recvsum);
+  std::size_t _recv_elems = rng::size(vec_recvdata);
+
+  default_comm().alltoallv(lsegment, vec_split_s, vec_split_i, vec_recvdata,
+                           vec_rsizes, vec_rindices);
+
+  __detail::local_sort(vec_recvdata, comp);
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:149]: As vec_recvdata is allocated in card memory, it is better to use fast dpl sort on GPU that do manual merge on CPU.

[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/mhp/algorithms/sort.hpp:149]: the best is to use manual merge on GPU

File: include/dr/mhp/algorithms/sort.hpp [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+                         default_comm().prev(), t, &req_l);
+    MPI_Wait(&req_l, &stat_l);
+
+  } else if (static_cast<int>(rng::size(vec_recvdata)) < -shift_right) {
+
+    assert(shift_left > 0);
+    default_comm().irecv(vec_left, default_comm().prev(), t, &req_l);
+    MPI_Wait(&req_l, &stat_l);
+
+    vec_recvdata.insert(vec_recvdata.begin(), vec_left.begin(), vec_left.end());
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:201]: If both vectors are allocated with the same allocator, they can be swapped. Inserting at the end is efficient enough for now (optimization stage will be next)

File: include/dr/mhp/algorithms/sort.hpp [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+
+  MPI_Request req_l, req_r;
+  MPI_Status stat_l, stat_r;
+  communicator::tag t = communicator::tag::halo_index;
+
+  std::vector<valT> vec_left(shift_left > 0 ? (shift_left) : 0);
+  std::vector<valT> vec_right(shift_right > 0 ? (shift_right) : 0);
+
+  if (static_cast<int>(rng::size(vec_recvdata)) < -shift_left) {
+
```
[the remark this answers is not in the document]
Comment (mateuszpn) [reply include/dr/mhp/algorithms/sort.hpp:182]: This condition will never occur. Values equal to splitter go to the left, and we may have an excess of elements to the left, but not a shortage.

File: include/dr/mhp/algorithms/sort.hpp [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+  std::vector<std::size_t> vec_rsizes(_comm_size, 0);
+  std::vector<std::size_t> vec_rindices(_comm_size, 0); // recv buffers
+
+  default_comm().alltoall(vec_split_s, vec_rsizes, 1);
+
+  std::exclusive_scan(vec_rsizes.begin(), vec_rsizes.end(),
+                      vec_rindices.begin(), 0);
+
+  const std::size_t _recv_elems =
+      std::reduce(vec_rsizes.begin(), vec_rsizes.end());
```
Comment (lslusarczyk) [inline include/dr/mhp/algorithms/sort.hpp:117]: `_recv_elems == vec_rindices.back() + vec_rsizes.back()`

File: include/dr/mhp/algorithms/sort.hpp [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+#endif
+    } else {
+      drlog.debug("cpu rng::sort, size {}\n", rng::size(r));
+      rng::sort(rng::begin(r), rng::end(r), comp);
+    }
+  }
+}
+
+template <dr::distributed_range R, typename Compare>
+void dist_sort(R &r, Compare &&comp) {
```
Comment (lslusarczyk) [inline include/dr/mhp/algorithms/sort.hpp:51]: this function is 240 lines long
Hard to read, e.g. because each local variable once defined is possibly used in all later code.
Please leave here TODO and create a task for refactoring this function into logical subfunctions once we fix more urgent issues like MPI bug and sort performance.

**Review by lslusarczyk — COMMENTED** [review]
Comments to recent changes.

**Review by lslusarczyk — APPROVED** [review]
